### PR TITLE
Analyze table function arguments

### PR DESCRIPTION
**Pull request by kasiafi**
This PR adds the Analyzer support for table and descriptor arguments of table functions.
These arguments are not yet supported in further phases of query processing, and they are caught in `RelationPlanner`.

No docs required. We should document how table and descriptor arguments work when we add full support for them.

Some SPI changes are introduced. However, the affected classes are marked as experimental, so probably we don't need to announce the changes in the release notes. Also, the changed classes can't be effectively used, as the related argument types are not supported beyond the Analyzer.

Beside the SPI changes mentioned above, no release notes required.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -1701,7 +1701,7 @@ else if (argument.getValue() instanceof Expression) {
                 }
                 Expression expression = (Expression) argument.getValue();
                 // 'descriptor' as a function name is not allowed in this context
-                if (argument.getValue() instanceof FunctionCall && ((FunctionCall) argument.getValue()).getName().hasSuffix(QualifiedName.of("descriptor"))) { // function name is always compared case-insensitive
+                if (expression instanceof FunctionCall && ((FunctionCall) expression).getName().hasSuffix(QualifiedName.of("descriptor"))) { // function name is always compared case-insensitive
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:1692]: This is the opposite of inlining a variable :)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 2]
```diff
@@ -235,6 +236,8 @@
 
     private Optional<TableExecuteHandle> tableExecuteHandle = Optional.empty();
 
+    // names of tables and aliased relations. All parts of a QualifiedName are lower-cased.
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:239]: Why are they lower cased? What happens with delimited names? That would seem to go against what the SQL spec describes.

File: core/trino-spi/src/main/java/io/trino/spi/ptf/DescriptorArgument.java [hunk 3]
```diff
@@ -33,20 +33,41 @@
         extends Argument
 {
     public static final DescriptorArgument NULL_DESCRIPTOR = builder().build();
-    private final Optional<Descriptor> descriptor;
+
+    @Nullable
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/ptf/DescriptorArgument.java:37]: Why did you change this Optional to a nullable field?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -136,7 +136,6 @@
 import io.trino.sql.tree.Delete;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:140]: Can you point out where is the name conflict? I can't see it.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 5]
```diff
@@ -2041,24 +2052,188 @@ public Optional<Integer> getAtMost()
...
+        private final String argumentName;
+        private final Optional<QualifiedName> name;
+        private final Relation relation;
+        private final Optional<List<Expression>> partitionBy; // it is allowed to partition by empty list
+        private final Optional<OrderBy> orderBy;
+        private final boolean pruneWhenEmpty;
+        private final boolean rowSemantics;
+        private final boolean passThroughColumns;
+
+        private TableArgumentAnalysis(String argumentName, Optional<QualifiedName> name, Relation relation, Optional<List<Expression>> partitionBy, Optional<OrderBy> orderBy, boolean pruneWhenEmpty, boolean rowSemantics, boolean passThroughColumns)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:2066]: Place arguments on separate lines for readability

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -1552,6 +1562,7 @@ protected Scope visitTableFunctionInvocation(TableFunctionInvocation node, Optio
             }
 
             // currently we don't support input tables, so the output consists of proper columns only
+            // TODO spec p. 123
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:1565]: Add more details -- at least mention the section title. "p. 123" is too vague (which document?) and subject to change as the document is updated

**Review by martint — COMMENTED** [review]
(still reviewing the last commit)

File: core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java [hunk 7]
```diff
@@ -5904,6 +5919,320 @@ public void testJsonArrayInAggregationContext()
...
+    public void testCopartitioning()
+    {
+        // TABLE(t1) is matched by fully qualified name: tpch.s1.t1. It matches the second copartition item s1.t1.
+        // Aliased relation TABLE(SELECT 1, 2) t1(x, y) is matched by unqualified name. It matches the first copartition item t1.
+        analyze("""
+                      SELECT * FROM TABLE(system.two_table_arguments_function(
+                                                                               input1 => TABLE(t1) PARTITION BY (a, b),
+                                                                               input2 => TABLE(SELECT 1, 2) t1(x, y) PARTITION BY (x, y)
+                                                                               COPARTITION (t1, s1.t1)))
+                """);
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java:6120]: There's a lot of unnecessary space:

```java
analyze("""
        SELECT * FROM TABLE(system.two_table_arguments_function(
            input1 => TABLE(t1) PARTITION BY (a, b),
            input2 => TABLE(SELECT 1, 2) t1(x, y) PARTITION BY (x, y)
            COPARTITION (t1, s1.t1)))
        """);
```

Note that the contents of the text block should be lined up with the the closing `"""`, otherwise the string will contain leading space at the beginning of each line:

For example:

```java
System.out.println("""
        a
        """);
```

prints:

```
a
```

while:


```java
System.out.println("""
        a
    """);
```

prints:

```
    a
```

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:1693]: Indeed. Renamed to "Use variable".

File: core/trino-spi/src/main/java/io/trino/spi/ptf/DescriptorArgument.java (hunk 3 above)
Comment (kasiafi) [reply core/trino-spi/src/main/java/io/trino/spi/ptf/DescriptorArgument.java]: It was meant to simplify the code. Reverted.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:140]: It happened in the `StatementAnalyzer` where we analyze the incoming `io.trino.sql.tree.TableArgument`, and use the `io.trino.spi.ptf.TableArgument.Builder`. I renamed the former (tree). Analogically with the descriptor arguments.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -1637,10 +1644,10 @@ private Map<String, Argument> analyzeArguments(Node node, List<ArgumentSpecifica
                 }
             }
 
-            return passedArguments.buildOrThrow();
+            return new TableFunctionArgumentsAnalysis(passedArguments.buildOrThrow(), tableArgumentAnalyses.build());
         }
 
-        private Argument analyzeArgument(ArgumentSpecification argumentSpecification, TableFunctionArgument argument)
+        private Argument analyzeArgument(ArgumentSpecification argumentSpecification, TableFunctionArgument argument, Optional<Scope> scope, ImmutableList.Builder<TableArgumentAnalysis> tableArgumentAnalyses)
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:1651]: I added a helper class to return both the `Argument`, and the optional `TableArgumentAnalysis`. `TableArgumentAnalysis` couldn't be extended to carry the `Argument`, because it only applies to table arguments, while `Argument` is created for each argument regardless of its type.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java (hunk 2 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:239]: The comment is not precise. It should say "All names are resolved case-insensitive".

I am aware that this is not compliant with the SQL spec. Regarding resolution of co-partitioned tables, I chose to follow how table references are resolved in Trino.

Currently in Trino, all table references, including CTEs and aliased relations, are resolved case-insensitive, and sadly, the 'delimited' property is disregarded. In theory, we could do spec-compliant resolution for the co-partitioned tables, but I think it would be unintuitive to the user. Similarly, we decided to resolve TF names in the usual non-ANSI way to avoid confusion between different function types.

If we decided to resolve co-partitioned table references in the spec-compliant way, we'd have to decide how "permissive" we want to be with respect to schema and catalog names.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 9]
```diff
@@ -2058,7 +2058,7 @@ public TableFunctionInvocationAnalysis(
         {
             this.catalogHandle = requireNonNull(catalogHandle, "catalogHandle is null");
             this.functionName = requireNonNull(functionName, "functionName is null");
-            this.arguments = requireNonNull(arguments, "arguments is null");
+            this.arguments = ImmutableMap.copyOf(requireNonNull(arguments, "arguments is null"));
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:2209]: requireNonNull not needed, since `ImmutableMap.copyOf` will not work with a null argument.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 10]
```diff
@@ -2041,24 +2041,174 @@ public Optional<Integer> getAtMost()
...
+                List<TableArgumentAnalysis> tableArgumentAnalyses,
+                List<List<String>> copartitioningLists,
                 ConnectorTableFunctionHandle connectorTableFunctionHandle,
                 ConnectorTransactionHandle transactionHandle)
         {
             this.catalogHandle = requireNonNull(catalogHandle, "catalogHandle is null");
             this.functionName = requireNonNull(functionName, "functionName is null");
             this.arguments = ImmutableMap.copyOf(requireNonNull(arguments, "arguments is null"));
+            this.tableArgumentAnalyses = ImmutableList.copyOf(requireNonNull(tableArgumentAnalyses, "tableArgumentAnalyses is null"));
+            this.copartitioningLists = ImmutableList.copyOf(requireNonNull(copartitioningLists, "copartitioningLists is null"));
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:2211]: requireNonNull not needed

**Review by martint — APPROVED** [review]
